File: ios/BiblosSurferTests/Services/LibraryServiceTests.swift
//
//  LibraryServiceTests.swift
//  BiblosSurferTests
//

import XCTest
@testable import BiblosSurfer

final class LibraryServiceTests: XCTestCase {
    private var directory: URL!
    private var covers: URL!
    private var service: LibraryService!

    override func setUpWithError() throws {
        directory = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
        covers = directory.appendingPathComponent("Covers", isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        service = LibraryService(
            booksDirectory: directory,
            coversDirectory: covers,
            bundledBookURLs: [TestFixtures.sampleBookURL].compactMap { $0 }
        )
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: directory)
    }

    func testBundledBookAppearsAlongsideUserBooks() async throws {
        let leftover = directory.appendingPathComponent("The Sample Voyage.epub")
        XCTAssertTrue(FileManager.default.createFile(atPath: leftover.path, contents: Data("old-mock".utf8)))

        let items = try await service.loadItems()
        let titles = Set(items.map(\.title))

        XCTAssertTrue(FileManager.default.fileExists(atPath: leftover.path))
        XCTAssertTrue(titles.contains(TestFixtures.sampleBookTitle))
        XCTAssertTrue(titles.contains("The Sample Voyage"))
        XCTAssertEqual(
            items.first { $0.title == TestFixtures.sampleBookTitle }?.fileURL,
            TestFixtures.sampleBookURL
        )
    }

    func testBundledBookIsListedFromTheBundleWithoutCopying() async throws {
        let items = try await service.loadItems()

        XCTAssertEqual(items.count, 1)
        XCTAssertEqual(items.first?.title, TestFixtures.sampleBookTitle)
        XCTAssertEqual(items.first?.format, .epub)
        XCTAssertEqual(items.first?.isEPUB, true)
        XCTAssertEqual(items.first?.fileURL, TestFixtures.sampleBookURL)
        let copiedBooks = try FileManager.default.contentsOfDirectory(at: directory, includingPropertiesForKeys: nil)
            .filter { ["epub", "pdf"].contains($0.pathExtension.lowercased()) }
        XCTAssertEqual(copiedBooks.count, 0)
    }

    func testNonBookFilesAreNotListed() async throws {
        _ = try await service.loadItems()
        let notes = directory.appendingPathComponent("notes.txt")
        XCTAssertTrue(FileManager.default.createFile(atPath: notes.path, contents: Data("hello".utf8)))

        let items = try await service.loadItems()

        XCTAssertEqual(items.map(\.title), [TestFixtures.sampleBookTitle])
    }

    func testImportingUnsupportedFormatThrows() async {
        let notes = directory.appendingPathComponent("notes.txt")
        XCTAssertTrue(FileManager.default.createFile(atPath: notes.path, contents: Data()))

        do {
            _ = try await service.importBook(from: notes)
            XCTFail("Expected unsupportedFormat")
        } catch let error as Errors.Library {
            guard case .unsupportedFormat(let fileExtension) = error else {
                return XCTFail("Expected unsupportedFormat, got \(error)")
            }
            XCTAssertEqual(fileExtension, "txt")
        } catch {
            XCTFail("Expected Errors.Library, got \(error)")
        }
    }

    func testImportingEPUBCopiesIntoTheLibrary() async throws {
        _ = try await service.loadItems()
        let source = try TestFixtures.temporarySampleBookURL()
        defer { try? FileManager.default.removeItem(at: source.deletingLastPathComponent()) }

        let imported = try await service.importBook(from: source)

        XCTAssertTrue(imported.isEPUB)
        XCTAssertTrue(imported.fileURL.path.hasPrefix(directory.path))
        XCTAssertTrue(FileManager.default.fileExists(atPath: imported.fileURL.path))

        let items = try await service.loadItems()
        XCTAssertEqual(items.count, 1)
        XCTAssertTrue(items.contains { $0.fileURL.path.hasPrefix(directory.path) })
    }
}

File: android/app/src/test/java/miko/biblossurfer/data/LibraryServiceTests.kt
package miko.biblossurfer.data

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.test.runTest
import miko.biblossurfer.support.TestFixtures
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.annotation.Config
import java.io.File
import java.util.UUID

@RunWith(AndroidJUnit4::class)
@Config(sdk = [33])
class LibraryServiceTests {
    private lateinit var directory: File
    private lateinit var covers: File
    private lateinit var service: LibraryService

    @Before
    fun setUp() {
        val context = ApplicationProvider.getApplicationContext<android.content.Context>()
        directory = File(context.cacheDir, UUID.randomUUID().toString())
        covers = File(directory, "Covers")
        directory.mkdirs()
        service = LibraryService(
            booksDirectory = directory,
            coversDirectory = covers,
            bundledBookFiles = listOfNotNull(TestFixtures.sampleBookFile),
            opener = PublicationOpeningService(context),
            bookStore = LocalBookService(),
        )
    }

    @After
    fun tearDown() {
        directory.deleteRecursively()
    }

    @Test
    fun bundledBookAppearsAlongsideUserBooks() = runTest {
        val leftover = File(directory, "The Sample Voyage.epub")
        leftover.writeText("old-mock")

        val items = service.loadItems()
        val titles = items.map { it.title }.toSet()

        assertTrue(leftover.exists())
        assertTrue(titles.contains(TestFixtures.sampleBookTitle))
        assertTrue(titles.contains("The Sample Voyage"))
        assertEquals(
            TestFixtures.sampleBookFile?.absolutePath,
            items.first { it.title == TestFixtures.sampleBookTitle }.fileURL.absolutePath,
        )
    }

    @Test
    fun bundledBookIsListedFromTheBundleWithoutCopying() = runTest {
        val items = service.loadItems()
        assertEquals(1, items.size)
        assertEquals(TestFixtures.sampleBookTitle, items.first().title)
        assertEquals(miko.biblossurfer.data.model.PublicationFormat.EPUB, items.first().format)
        assertEquals(true, items.first().isEPUB)
        assertEquals(TestFixtures.sampleBookFile?.absolutePath, items.first().fileURL.absolutePath)
        val copiedBooks = directory.listFiles().orEmpty()
            .filter { it.extension.lowercase() in listOf("epub", "pdf") }
        assertEquals(0, copiedBooks.size)
    }

    @Test
    fun nonBookFilesAreNotListed() = runTest {
        service.loadItems()
        File(directory, "notes.txt").writeText("hello")
        val items = service.loadItems()
        assertEquals(listOf(TestFixtures.sampleBookTitle), items.map { it.title })
    }

    @Test
    fun importingUnsupportedFormatThrows() = runTest {
        val notes = File(directory, "notes.txt")
        notes.writeBytes(ByteArray(0))
        try {
            service.importBook(notes)
            fail("Expected unsupportedFormat")
        } catch (error: Errors.Library.UnsupportedFormat) {
            assertEquals("txt", error.fileExtension)
        } catch (error: Throwable) {
            fail("Expected Errors.Library.UnsupportedFormat, got $error")
        }
    }

    @Test
    fun importingEPUBCopiesIntoTheLibrary() = runTest {
        service.loadItems()
        val source = TestFixtures.temporarySampleBookFile()
        try {
            val imported = service.importBook(source)
            assertTrue(imported.isEPUB)
            assertTrue(imported.fileURL.path.startsWith(directory.path))
            assertTrue(imported.fileURL.exists())
            val items = service.loadItems()
            assertEquals(1, items.size)
            assertTrue(items.any { it.fileURL.path.startsWith(directory.path) })
        } finally {
            source.parentFile?.deleteRecursively()
        }
    }
}
